File: web/src/models/fit.ts
import { useEffect, useState } from 'react'
import { invoke } from '@tauri-apps/api/core'

import { size } from '@/shell/bytes'

/**
 * The fit verdict, as the window holds it
 * ([#74](https://github.com/elpideus/demido-studio/issues/74)).
 *
 * `demido-vram` decides; the window writes the sentence. The verdict is read
 * **again while the pane is open** rather than once when it was drawn, because
 * the card's free memory moves by more than the rig's tightest row has to
 * spare whenever a browser window opens. And nothing that starts a download
 * reads it: the verdict informs, and the machine is the person's.
 */

/** The Rust `demido_vram::Verdict`. Every figure is in bytes. */
export type Verdict =
  | { fit: 'fits'; needs: number; room: number; spare: number; contextPriced: boolean }
  | { fit: 'partial'; needs: number; room: number; over: number; contextPriced: boolean }
  | { fit: 'unread'; needs: number }
  | { fit: 'unpriced' }

/** How often an open pane reads the card again. */
const AGAIN_EVERY_MS = 5000

/** Whether this card can hold weights of `weights` bytes, read now. */
export function readFit(weights: number): Promise<Verdict> {
  return invoke<Verdict>('models_fit', { weights })
}

/**
 * The verdict for weights of `weights` bytes, read now, again every few
 * seconds while mounted, and again when the window comes back into focus.
 * `undefined` until the first reading answers, or when it could not be asked.
 */
export function useFit(weights: number): Verdict | undefined {
  const [verdict, setVerdict] = useState<Verdict>()

  useEffect(() => {
    let live = true
    const read = () => {
      readFit(weights).then(
        (read) => live && setVerdict(read),
        () => live && setVerdict(undefined),
      )
    }
    setVerdict(undefined)
    read()
    const every = window.setInterval(read, AGAIN_EVERY_MS)
    window.addEventListener('focus', read)
    return () => {
      live = false
      window.clearInterval(every)
      window.removeEventListener('focus', read)
    }
  }, [weights])

  return verdict
}

/** The verdict in two parts: a label a person scans, and the sentence behind
 * it. Each of the four is something different to do. */
export function words(verdict: Verdict): { label: string; sentence: string } {
  switch (verdict.fit) {
    case 'fits':
      return {
        label: 'Fits on this card',
        sentence: verdict.contextPriced
          ? `${size(verdict.needs)} leaves ${size(verdict.spare)} of the ${size(verdict.room)} this card has free.`
          : `The weights take ${size(verdict.needs)} of the ${size(verdict.room)} this card has free, before the context.`,
      }
    case 'partial':
      return {
        label: 'Partial offload',
        sentence: `${size(verdict.over)} over the ${size(verdict.room)} this card has free, so part of it runs from system memory, more slowly.`,
      }
    case 'unread':
      return {
        label: 'Card not read',
        sentence: `The weights take ${size(verdict.needs)}. This card's free memory could not be read.`,
      }
    case 'unpriced':
      return {
        label: 'Size not published',
        sentence: 'The server states no size for these weights, so there is nothing to price.',
      }
  }
}

File: src-tauri/src/fit.rs
//! The fit verdict, as the detail pane asks for it
//! ([#74](https://github.com/elpideus/demido-studio/issues/74)).
//!
//! `demido-vram` owns the reading and the arithmetic, and the conversation
//! knows what its resident model was weighed at holding, which a load would
//! give back (`demido_chat::Chat::held`). This is where the two meet.
//!
//! **The card is read every time the pane asks.** Its free memory is not what
//! it was when the pane was drawn: a browser window opened in between moves it
//! by more than `done.md`'s tightest row has to spare. So nothing here keeps a
//! reading, the window asks again while the pane is open, and the download
//! never acts on a verdict at all: `downloads_add` takes a repository and a
//! choice, and no verdict is among its arguments. A verdict informs; it cannot
//! be the thing a stale reading turned into a refusal.

use std::path::PathBuf;

use demido_vram::{free_now, verdict, Load, Verdict, BESIDE_THE_KV};

use crate::wiring::Wiring;

/// Whether this card can hold weights of `weights` bytes, read now.
///
/// `path` is the model's file when it is already on disk, and then the context
/// is priced too: one slot at the context length in force, read from the
/// file's header by the same reading that prices a slot for the pool, with
/// what the build holds beside it
/// (`demido_models::slot`,
/// [#105](https://github.com/elpideus/demido-studio/issues/105)). A file not
/// yet downloaded has no header to read, and an architecture the reading does
/// not know is not priced, so for both the verdict says *before the context*
/// rather than adding a guess.
#[tauri::command]
pub async fn models_fit(
    wiring: tauri::State<'_, Wiring>,
    weights: u64,
    path: Option<PathBuf>,
) -> demido_core::Result<Verdict> {
    let context = match path {
        Some(path) => {
            let tokens = wiring.chat.resolved().context_length();
            tauri::async_runtime::spawn_blocking(move || {
                demido_models::price(&path, tokens).per_slot
            })
            .await
            .ok()
            .flatten()
            // What the build holds beside the KV, the way the pool counts it,
            // so the verdict and the pool agree about one load.
            .map(|kv| kv.saturating_add(BESIDE_THE_KV))
        }
        None => None,
    };
    Ok(verdict(
        free_now(),
        wiring.chat.held().await,
        Load { weights, context },
    ))
}
